### PR TITLE
Expect path to wallet file, not dir. Use a cache for account addresses.

### PR DESCRIPTION
**Pull request by mitchmindtree**
Closes https://github.com/FuelLabs/forc-wallet/issues/78, taking the approach mentioned here:

https://github.com/FuelLabs/forc-wallet/issues/78#issuecomment-1425160146

When a path to a wallet is specified, we now expect a path to a keystore file. The file can be anywhere on the user's machine and no longer requires a `.accounts` file in the same directory to track derived accounts.

Instead, we now track derived accounts within `~/.fuel/wallets/accounts`, where each wallet has its own unique cache directory named by hashing the wallet's ciphertext into a fixed-length hex string. Each directory contains a file for each derived account address, where the file name is
the wallet's account index and the file itself contains the address as a utf8 string.

This PR also removes the `serialtest` dev dependency in favour of generating unique directory names for temporary test directories.

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -193,26 +170,6 @@ mod tests {
...
-    fn handle_none_argument() {
-        let path_opt: Option<PathBuf> = None;
-        let path = path_opt.unwrap_or_else(default_wallet_path);
-        validate_wallet_path(&path).unwrap();
-        let default_path = default_wallet_path();
-        assert_eq!(path, default_path)
-    }
-
-    #[test]
-    fn handle_relative_path_argument() {
```
Comment (sdankel) [inline src/utils.rs:206]: Does the CLI still support relative paths?

**Review by sdankel — APPROVED** [review]

File: src/account.rs [hunk 2]
```diff
@@ -1,69 +1,147 @@
...
     } else {
-        eprintln!("Account {account_index} is not derived yet!");
+        eprintln!("Account {account_ix} is not derived yet!");
     }
     Ok(())
 }
 
-fn new_account(wallet_dir: &Path, password: &str) -> Result<WalletUnlocked> {
-    let account_index = number_of_derived_accounts(wallet_dir);
+fn new_account(wallet_path: &Path, password: &str) -> Result<WalletUnlocked> {
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply src/account.rs:35]: `?` will only work for expressions where the expression has some type that can be unambiguously converted into the function's return type - in this case, our return type is `anyhow::Result<T>` (short for `Result<T, anyhow::Error>`). There's no 

Here's the magical trait that enables "short-circuiting" expressions with the `?` operator: https://doc.rust-lang.org/std/ops/trait.Try.html. Note that `Option` can only be `Try`'d and return early with an `Option`, similarly `Result` can only return early into another `Result`. So in our case here, the returned `anyhow::Error` will always come from some other `Error` instance produced by the `?` expression.

Moral of the story: we can't actually return via `?` on a `None` here :sweat_smile:  Is that kind of what you were asking?

File: src/account.rs [hunk 3]
```diff
@@ -1,69 +1,147 @@
...
-    #[serial]
     fn create_new_account() {
         with_tmp_folder(|tmp_folder| {
             // init test wallet
-            save_dummy_wallet_file(tmp_folder);
-            let account_is_ok = new_account(tmp_folder, TEST_PASSWORD).is_ok();
-            assert!(account_is_ok)
+            let wallet_path = tmp_folder.join("wallet.json");
+            save_dummy_wallet_file(&wallet_path);
+            new_account(&wallet_path, TEST_PASSWORD).unwrap();
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply src/account.rs:144]: > Do we need to assert that this doesn't throw an error, or will that cause the test to fail?

Yes calling `.unwrap()` on a `Result` will always fail if the `Result` is `Err`.

> Might be good to have another test case where it throws.

Ah you mean have another test where we make sure we fail if the password's wrong? Sounds good!

File: src/utils.rs [hunk 4]
```diff
@@ -1,58 +1,58 @@
...
-const USER_FUEL_DIR: &str = ".fuel";
-/// The directory under which `forc wallet` generates wallets.
-const WALLETS_DIR: &str = "wallets";
-/// The file stem for the wallet file, a JSON file with AES encrypted keys etc.
-const WALLET_FILE_STEM: &str = ".wallet";
-/// The file storing wallet account information.
-const ACCOUNTS_FILE_STEM: &str = ".accounts";
+/// The user's fuel directory (stores state related to fuel-core, wallet, etc).
+pub fn user_fuel_dir() -> PathBuf {
+    const USER_FUEL_DIR: &str = ".fuel";
```
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply src/utils.rs:16]: I decided to move these into the functions to clarify these are the only locations they're actually needed - I figure we can pull them out into the module-level if we need to share them, but the idea is that the user should probably use one of these helper functions to construct the path they need anyways.

File: src/utils.rs (hunk 1 above)
Comment (mitchmindtree) [reply src/utils.rs:206]: Ahh good catch! I'll re-add this. I think I removed this thinking it was already tested elsewhere, but it doesn't look like it is.

**Pull request merged**